requests_ntlm/requests_ntlm.py

```diff
@@ -180,3 +178,32 @@ def _auth_type_from_header(header):
         return 'Negotiate'
 
     return None
+
+def _get_server_cert(request_url):
+    """
+    Get the certificate at the request_url and return it as a SHA256 hash. Will check the endpoint if it
+    is a https site and use the default port 443 is it isn't explicitly specified in the URL. Used to send
+    in with NTLMv2 authentication for Channel Binding Tokens
+
+    :param request_url: The request url in the format https://endpoint:port/path
+    :return: SHA256 hash of the DER encoded certificate at the request_url or None if not a HTTPS endpoint
+    """
+    host_pattern = re.compile('(?i)^https://?(?P<host>[0-9a-z-_.]+)(:(?P<port>\d+))?')
+    match = host_pattern.match(request_url)
+
+    if match:
+        host = match.group('host')
+        port = match.group('port')
+        if not port:
+            port = 443
+        else:
+            port = int(port)
+
+        s = socket()
+        c = ssl.wrap_socket(s)
+        c.connect((host, port))
+        server_certificate = c.getpeercert(True)
+        hash_object = hashlib.sha256(server_certificate)
+        return hash_object.hexdigest().upper()
```

Comment:
@jborean93 So by default requests calls into urllib3, and gets back a urllib3 `HTTPResponse` object that ends up in `Response.raw`. However, there is no *requirement* that that be the call, and frequently users try a different backend or put a different file object there (e.g. `BytesIO`). For that reason, Requests core projects must not unconditionally assume that `Response.raw` is a urllib3 `HTTPResponse` object. They may attempt to use it like one, but if they fail they should fall back to a sensible behaviour.

So I'm saying we should look in those places first, but if those places don't exist we should tolerate that rather than throwing exceptions.